nixos/modules/virtualisation/xen-dom0.nix

```diff
@@ -1,213 +1,739 @@
+        };
       };
 
-    virtualisation.xen.stored =
-      mkOption {
-        type = types.path;
-        description =
-          ''
-            Xen Store daemon to use. Defaults to oxenstored of the xen package.
-          '';
+      bootParams = lib.options.mkOption {
+        default = [ ];
+        example = ''
+          [
+            "iommu=force:true,qinval:true,debug:true"
+            "noreboot=true"
+            "vga=ask"
+          ]
+        '';
+        type = lib.types.listOf lib.types.str;
+        description = ''
+          Parameters passed to the Xen Hypervisor at boot time.
+        '';
       };
 
-    virtualisation.xen.domains = {
-        extraConfig = mkOption {
-          type = types.lines;
+      domains = {
+        extraConfig = lib.options.mkOption {
+          type = lib.types.lines;
           default = "";
-          description =
-            ''
-              Options defined here will override the defaults for xendomains.
-              The default options can be seen in the file included from
-              /etc/default/xendomains.
-            '';
-          };
+          example = ''
+            XENDOMAINS_SAVE=/persist/xen/save
+            XENDOMAINS_RESTORE=false
+            XENDOMAINS_CREATE_USLEEP=10000000
+          '';
+          description = ''
+            Options defined here will override the defaults for xendomains.
+            The default options can be seen in the file included from
+            /etc/default/xendomains.
+          '';
+        };
       };
-
-    virtualisation.xen.trace = mkEnableOption "Xen tracing";
-
+    };
   };
 
-
   ###### implementation
 
-  config = mkIf cfg.enable {
-    assertions = [ {
-      assertion = pkgs.stdenv.isx86_64;
-      message = "Xen currently not supported on ${pkgs.stdenv.hostPlatform.system}";
-    } {
-      assertion = config.boot.loader.grub.enable && (config.boot.loader.grub.efiSupport == false);
-      message = "Xen currently does not support EFI boot";
-    } ];
-
-    virtualisation.xen.package = mkDefault pkgs.xen;
-    virtualisation.xen.package-qemu = mkDefault pkgs.xen;
-    virtualisation.xen.stored = mkDefault "${cfg.package}/bin/oxenstored";
-
-    environment.systemPackages = [ cfg.package ];
-
-    boot.kernelModules =
-      [ "xen-evtchn" "xen-gntdev" "xen-gntalloc" "xen-blkback" "xen-netback"
-        "xen-pciback" "evtchn" "gntdev" "netbk" "blkbk" "xen-scsibk"
-        "usbbk" "pciback" "xen-acpi-processor" "blktap2" "tun" "netxen_nic"
-        "xen_wdt" "xen-acpi-processor" "xen-privcmd" "xen-scsiback"
+  config = lib.modules.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = pkgs.stdenv.isx86_64;
+        message = "Xen is currently not supported on ${pkgs.stdenv.hostPlatform.system}";
+      }
+      {
+        assertion = cfg.efi.enable == config.boot.loader.systemd-boot.enable;
+        message = "Xen currently does not support EFI boot on bootloaders that are not systemd-boot";
```

Comment:
This fails to evaluate with [lanzaboote](https://github.com/nix-community/lanzaboote) provide secure-boot capability for `systemd-boot`.

Refer to NiXium's attempted deployment at https://github.com/NiXium-org/NiXium/pull/121#issuecomment-2249090074

Reply:
Yes. Lanzaboote doesn't "provide secure-boot capability for systemd-boot"; it's an entirely new UEFI stub, and it requires patching to work with Xen's edge case.

Aside from that, Secure Boot for Xen in general is a whole other [can](https://www.youtube.com/watch?v=A_IhKjK7EgA) of [worms](https://github.com/tklengyel/xen-uefi) - one that is, in practice, extremely difficult to implement. If I do get to work on securing Xen's boot process, it probably won't be through Secure Boot, but through [Trenchboot](https://trenchboot.org/) or some other trusted boot process.

Reply:
:thinking: Noted

Reply:
@SigmaSquadron Can you elaborate how would you do that through the trenchboot? Or like my concern is managing the threat of someone injecting unauthorized kernel and OS on the system, is there a coreboot (NiXium is migrating on coreboot x86 systems atm where the candidate system has coreboot support already) option that you could recommend?

Reply:
I can't, because I have no idea, and Trenchboot might as well be a fantasy at this point in time. Your best bet, if you really need it, is to start the work on Xen support for Lanzaboote yourself, and see if signing the EFI just magically works.